File: server/versionManager.js
/**
 * server/versionManager.js
 * Handles project versioning operations:
 *   - archiveVersion()    snapshot current project into versions/
 *   - compressVersions()  gzip any loose version dirs in versions/
 *   - applyUpdate()       extract a tarball over the project, archive first
 *   - listVersions()      list all versions with metadata
 */

import fs   from 'node:fs';
import path from 'node:path';
import zlib from 'node:zlib';
import { loadProjectConfig, saveProjectConfig, bumpVersion } from '../parser/projectConfig.js';

// ─── Tar writer (pure Node.js) ────────────────────────────────────────────────

function encodeOctal(n, len) {
  return n.toString(8).padStart(len - 1, '0') + '\0';
}

function makeTarHeader(filePath, size, mtime, isDir) {
  const buf = Buffer.alloc(512, 0);
  const name = filePath.slice(0, 100);
  buf.write(name, 0, 'utf8');                             // name
  buf.write(encodeOctal(isDir ? 0o755 : 0o644, 8), 100); // mode
  buf.write(encodeOctal(0, 8), 108);                      // uid
  buf.write(encodeOctal(0, 8), 116);                      // gid
  buf.write(encodeOctal(size, 12), 124);                  // size
  buf.write(encodeOctal(Math.floor(mtime / 1000), 12), 136); // mtime
  buf.write('        ', 148);                             // checksum placeholder
  buf[156] = isDir ? 0x35 : 0x30;                        // typeflag: '5' dir, '0' file
  buf.write('ustar\0', 257);                              // magic
  buf.write('00', 263);                                   // version
  // Compute checksum
  let sum = 0;
  for (let i = 0; i < 512; i++) sum += buf[i];
  buf.write(encodeOctal(sum, 7) + ' ', 148);
  return buf;
}

function tarDirectory(srcDir, tarStream) {
  const BLOCK = 512;

  function addEntry(absPath, relPath) {
    const stat = fs.statSync(absPath);
    if (stat.isDirectory()) {
      tarStream.push(makeTarHeader(relPath + '/', 0, stat.mtimeMs, true));
      for (const entry of fs.readdirSync(absPath).sort()) {
        addEntry(path.join(absPath, entry), relPath + '/' + entry);
      }
    } else {
      const data = fs.readFileSync(absPath);
      tarStream.push(makeTarHeader(relPath, data.length, stat.mtimeMs, false));
      // Data padded to 512-byte blocks
      const padded = Buffer.alloc(Math.ceil(data.length / BLOCK) * BLOCK, 0);
      data.copy(padded);
      tarStream.push(padded);
    }
  }

  const name = path.basename(srcDir);
  addEntry(srcDir, name);
  // End-of-archive: two 512-byte zero blocks
  tarStream.push(Buffer.alloc(1024, 0));
}

function dirToTarGz(srcDir, destFile) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    tarDirectory(srcDir, chunks);
    const tarBuf = Buffer.concat(chunks);
    zlib.gzip(tarBuf, (err, gz) => {
      if (err) return reject(err);
      fs.writeFileSync(destFile, gz);
      resolve(destFile);
    });
  });
}

// ─── Tar reader (for applying updates) ───────────────────────────────────────

function parseTar(tarBuf, outDir) {
  let off = 0;
  while (off + 512 <= tarBuf.length) {
    const hdr      = tarBuf.slice(off, off + 512);
    const name     = hdr.slice(0, 100).toString('utf8').replace(/\0/g, '').trim();
    if (!name) { off += 512; continue; }
    const size     = parseInt(hdr.slice(124, 136).toString().replace(/\0/g, '').trim(), 8) || 0;
    const typeflag = String.fromCharCode(hdr[156]);
    off += 512;
    // Sanitize path
    const safe = name.replace(/^[./\\]+/, '').replace(/\.\.\//g, '');
    // Strip top-level dir prefix if present (common in tarballs)
    const parts = safe.split('/');
    const stripped = parts.length > 1 && !parts[0].includes('.') ? parts.slice(1).join('/') : safe;
    if (stripped) {
      const dest = path.join(outDir, stripped);
      if (typeflag === '5') {
        fs.mkdirSync(dest, { recursive: true });
      } else if (typeflag === '0' || typeflag === '\0') {
        fs.mkdirSync(path.dirname(dest), { recursive: true });
        fs.writeFileSync(dest, tarBuf.slice(off, off + size));
      }
    }
    off += Math.ceil(size / 512) * 512;
  }
}

async function extractTarGz(gzBuf, destDir) {
  const tarBuf = await new Promise((res, rej) => {
    zlib.gunzip(gzBuf, (err, buf) => err ? rej(err) : res(buf));
  });
  parseTar(tarBuf, destDir);
}

// ─── Version operations ───────────────────────────────────────────────────────

/**
 * Copy/snapshot the current project into versions/<version>-<timestamp>/
 * Returns the path of the snapshot dir.
 */
export async function archiveVersion(projectRoot) {
  const config     = loadProjectConfig(projectRoot);
  const versionsDir = path.join(projectRoot, config.versions?.dir || 'versions');
  const version    = config.version || '0.0.0';
  const ts         = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  const snapName   = `v${version}-${ts}`;
  const snapDir    = path.join(versionsDir, snapName);

  fs.mkdirSync(versionsDir, { recursive: true });

  if (config.versions?.compress) {
    // Compress directly to .tar.gz
    const tarPath = snapDir + '.tar.gz';
    await dirToTarGz(projectRoot, tarPath);
    pruneOldVersions(versionsDir, config.versions?.keep ?? 20);
    return tarPath;
  } else {
    // Copy as loose directory
    copyDirSync(projectRoot, snapDir, [config.versions?.dir || 'versions']);
    pruneOldVersions(versionsDir, config.versions?.keep ?? 20);
    return snapDir;
  }
}

/**
 * Compress any loose version directories in versions/ to .tar.gz
 * and remove the originals. Returns list of compressed paths.
 */
export async function compressVersions(projectRoot) {
  const config      = loadProjectConfig(projectRoot);
  const versionsDir = path.join(projectRoot, config.versions?.dir || 'versions');
  if (!fs.existsSync(versionsDir)) return [];

  const entries = fs.readdirSync(versionsDir, { withFileTypes: true });
  const results = [];

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const dirPath = path.join(versionsDir, entry.name);
    const tarPath = dirPath + '.tar.gz';
    try {
      await dirToTarGz(dirPath, tarPath);
      fs.rmSync(dirPath, { recursive: true, force: true });
      results.push({ name: entry.name, tarball: tarPath });
    } catch(e) {
      results.push({ name: entry.name, error: e.message });
    }
  }
  return results;
}

/**
 * Apply an update tarball to a project:
 *   1. Archive current version first
 *   2. Extract the tarball over the project dir
 *   3. Bump version in side.project.json
 *   4. Return new version string
 */
export async function applyUpdate(projectRoot, gzBuf, bumpPart = 'patch') {
  // 1. Archive current state
  const archivePath = await archiveVersion(projectRoot);

  // 2. Extract new version over project
  await extractTarGz(gzBuf, projectRoot);

  // 3. Bump version
  const config     = loadProjectConfig(projectRoot);
  const newVersion = bumpVersion(config.version, bumpPart);
  config.version   = newVersion;
  saveProjectConfig(projectRoot, config);

  return { newVersion, archivePath };
}

/**
 * List all versions in the versions/ dir with size + date metadata.
 */
export function listVersions(projectRoot) {
  const config      = loadProjectConfig(projectRoot);
  const versionsDir = path.join(projectRoot, config.versions?.dir || 'versions');
  if (!fs.existsSync(versionsDir)) return [];

  return fs.readdirSync(versionsDir, { withFileTypes: true })
    .filter(e => e.isFile() && e.name.endsWith('.tar.gz') || e.isDirectory())
    .map(e => {
      const fullPath = path.join(versionsDir, e.name);
      const stat     = fs.statSync(fullPath);
      return {
        name:      e.name,
        type:      e.isDirectory() ? 'dir' : 'tarball',
        size:      stat.size,
        modified:  stat.mtime.toISOString(),
        path:      fullPath,
      };
    })
    .sort((a, b) => b.modified.localeCompare(a.modified));
}

// ─── Utilities ────────────────────────────────────────────────────────────────

function copyDirSync(src, dest, excludeDirs = []) {
  fs.mkdirSync(dest, { recursive: true });
  for (const entry of fs.readdirSync(src, { withFileTypes: true })) {
    if (entry.name.startsWith('.')) continue;
    if (excludeDirs.includes(entry.name)) continue;
    const s = path.join(src, entry.name);
    const d = path.join(dest, entry.name);
    if (entry.isDirectory()) copyDirSync(s, d, excludeDirs);
    else fs.copyFileSync(s, d);
  }
}

function pruneOldVersions(versionsDir, keep = 20) {
  if (!fs.existsSync(versionsDir)) return;
  const entries = fs.readdirSync(versionsDir, { withFileTypes: true })
    .map(e => ({
      name: e.name,
      path: path.join(versionsDir, e.name),
      mtime: fs.statSync(path.join(versionsDir, e.name)).mtimeMs,
    }))
    .sort((a, b) => b.mtime - a.mtime);

  for (const entry of entries.slice(keep)) {
    fs.rmSync(entry.path, { recursive: true, force: true });
  }
}

File: version/version_manager.py
# SPDX-License-Identifier: GPL-3.0-or-later
# Copyright (C) 2026 N0V4-N3XU5

"""
version/version_manager.py
==========================
Handles all project versioning operations using the stdlib tarfile module.

Operations
----------
archive_version(root)
    Snapshot the current project into versions/<v>-<timestamp>.tar.gz
    Respects the ignore patterns from side.project.json.
    Prunes old snapshots beyond the configured keep limit.

apply_update(root, tarball_path, bump_part)
    1. Archive current state first (safety net)
    2. Extract the tarball over the project directory
    3. Bump the version in side.project.json
    Returns (new_version, archive_path)

list_versions(root)
    Return metadata for all snapshots in the versions/ directory,
    sorted newest-first.

compress_loose(root)
    Find any uncompressed snapshot directories in versions/ and
    convert them to .tar.gz, removing the originals.

The tarball format is standard .tar.gz with the project directory
name as the top-level prefix, matching what most tools expect.
Path traversal is sanitised on extract (no ../ escapes).
"""

from __future__ import annotations
import os
import tarfile
import fnmatch
from datetime import datetime, timezone
from pathlib import Path

from parser.project_config import load_project_config, save_project_config, bump_version, init_project_config


# ── Ignore patterns applied during archive ────────────────────────────────────
# These supplement the project's own ignore list and prevent the archive
# from including version snapshots of version snapshots.
_ARCHIVE_EXCLUDE_ALWAYS = {
    "__pycache__", ".git", ".venv", "venv", "env",
    "node_modules", ".mypy_cache", ".pytest_cache",
    ".cache", "dist", "build",
}


def _make_filter(versions_dir_name: str, extra_ignore: list[str]):
    """
    Return a tarfile filter function that excludes:
      - the versions directory itself
      - always-excluded dirs
      - glob patterns from extra_ignore
    """
    def _filter(tarinfo: tarfile.TarInfo) -> tarfile.TarInfo | None:
        # tarinfo.name is like 'projectname/subdir/file'
        parts = Path(tarinfo.name).parts
        # Check each path component
        for part in parts:
            if part in _ARCHIVE_EXCLUDE_ALWAYS:
                return None
            if part == versions_dir_name:
                return None
            for pattern in extra_ignore:
                if fnmatch.fnmatch(part, pattern):
                    return None
        return tarinfo
    return _filter


def archive_version(root_dir: str) -> str:
    """
    Create a compressed snapshot of root_dir in its versions/ subdirectory.

    Returns the path of the created .tar.gz file.
    """
    root_dir = os.path.abspath(root_dir)
    # init_project_config creates side.project.json if it doesn't exist yet,
    # ensuring fresh projects get versioned correctly from their first archive
    config = init_project_config(root_dir)
    versions_subdir = (config.get("versions") or {}).get("dir", "versions")
    keep = int((config.get("versions") or {}).get("keep", 20))
    version = config.get("version") or "0.0.0"
    extra_ignore = list(config.get("ignore") or [])

    versions_dir = os.path.join(root_dir, versions_subdir)
    os.makedirs(versions_dir, exist_ok=True)

    ts = datetime.now(tz=timezone.utc).strftime("%Y%m%dT%H%M%S")
    archive_name = f"v{version}-{ts}.tar.gz"
    archive_path = os.path.join(versions_dir, archive_name)

    tar_filter = _make_filter(versions_subdir, extra_ignore)

    with tarfile.open(archive_path, "w:gz") as tar:
        # Add all files from root_dir, filtered
        project_name = os.path.basename(root_dir)
        for entry in sorted(os.listdir(root_dir)):
            full = os.path.join(root_dir, entry)
            arcname = os.path.join(project_name, entry)
            tar.add(full, arcname=arcname, recursive=True, filter=tar_filter)

    _prune_old_versions(versions_dir, keep)
    return archive_path


def apply_update(root_dir: str, tarball_path: str, bump_part: str = "patch") -> tuple[str, str]:
    """
    Apply a tarball update to root_dir:
      1. Archive current state
      2. Extract tarball over project (stripping top-level dir prefix)
      3. Bump version in side.project.json

    Returns (new_version, archive_path).
    Raises ValueError if tarball_path does not exist or is not a valid tar.
    """
    root_dir = os.path.abspath(root_dir)
    tarball_path = os.path.abspath(tarball_path)

    if not os.path.isfile(tarball_path):
        raise ValueError(f"Tarball not found: {tarball_path}")

    # 1. Safety archive
    archive_path = archive_version(root_dir)

    # 2. Extract (strip top-level directory component, sanitise paths)
    _extract_tarball(tarball_path, root_dir)

    # 3. Bump version
    config = load_project_config(root_dir)
    new_version = bump_version(config.get("version", "0.0.0"), bump_part)
    config["version"] = new_version
    save_project_config(root_dir, config)

    return new_version, archive_path


def list_versions(root_dir: str) -> list[dict]:
    """
    Return metadata for all snapshots in the versions/ directory,
    sorted newest-first.

    Each entry: {name, type, size, modified, path}
    """
    root_dir = os.path.abspath(root_dir)
    # init_project_config creates side.project.json if it doesn't exist yet,
    # ensuring fresh projects get versioned correctly from their first archive
    config = init_project_config(root_dir)
    versions_subdir = (config.get("versions") or {}).get("dir", "versions")
    versions_dir = os.path.join(root_dir, versions_subdir)

    if not os.path.isdir(versions_dir):
        return []

    results = []
    for entry in os.listdir(versions_dir):
        full = os.path.join(versions_dir, entry)
        try:
            stat = os.stat(full)
        except OSError:
            continue
        is_tarball = entry.endswith(".tar.gz") or entry.endswith(".tgz")
        is_dir = os.path.isdir(full)
        if not (is_tarball or is_dir):
            continue
        results.append({
            "name":     entry,
            "type":     "tarball" if is_tarball else "directory",
            "size":     stat.st_size,
            "modified": datetime.fromtimestamp(stat.st_mtime, tz=timezone.utc).isoformat(),
            "path":     full,
        })

    results.sort(key=lambda x: x["modified"], reverse=True)
    return results


def compress_loose(root_dir: str) -> list[dict]:
    """
    Find uncompressed snapshot directories in versions/ and convert them
    to .tar.gz archives, removing the originals.

    Returns list of {name, tarball, error} dicts.
    """
    root_dir = os.path.abspath(root_dir)
    # init_project_config creates side.project.json if it doesn't exist yet,
    # ensuring fresh projects get versioned correctly from their first archive
    config = init_project_config(root_dir)
    versions_subdir = (config.get("versions") or {}).get("dir", "versions")
    versions_dir = os.path.join(root_dir, versions_subdir)

    if not os.path.isdir(versions_dir):
        return []

    results = []
    for entry in os.listdir(versions_dir):
        full = os.path.join(versions_dir, entry)
        if not os.path.isdir(full):
            continue
        tarball_path = full + ".tar.gz"
        try:
            with tarfile.open(tarball_path, "w:gz") as tar:
                tar.add(full, arcname=entry)
            # Remove the loose directory after successful compression
            import shutil
            shutil.rmtree(full)
            results.append({"name": entry, "tarball": tarball_path})
        except Exception as exc:
            results.append({"name": entry, "error": str(exc)})

    return results


# ── Internal helpers ──────────────────────────────────────────────────────────

def _extract_tarball(tarball_path: str, dest_dir: str) -> None:
    """
    Extract a .tar.gz into dest_dir, stripping the top-level directory
    prefix that most project tarballs include (e.g. 'myproject/src/...'
    becomes 'src/...' in dest_dir).

    Sanitises member paths to prevent path-traversal attacks.
    """
    with tarfile.open(tarball_path, "r:gz") as tar:
        members = tar.getmembers()

        # Detect common top-level prefix (e.g. all members start with 'myproject/')
        prefix = _detect_prefix(members)

        for member in members:
            # Strip prefix
            rel = member.name
            if prefix and rel.startswith(prefix):
                rel = rel[len(prefix):]
            # Sanitise: no absolute paths, no ../ traversal
            rel = rel.lstrip("/\\")
            rel = os.path.normpath(rel)
            if rel.startswith(".."):
                continue   # skip anything that would escape dest_dir
            if not rel or rel == ".":
                continue

            dest_path = os.path.join(dest_dir, rel)

            if member.isdir():
                os.makedirs(dest_path, exist_ok=True)
            elif member.isfile():
                os.makedirs(os.path.dirname(dest_path), exist_ok=True)
                with tar.extractfile(member) as src, open(dest_path, "wb") as dst:
                    dst.write(src.read())


def _detect_prefix(members: list[tarfile.TarInfo]) -> str:
    """
    If all members share a common first path component that contains no '.'
    (i.e. looks like a project directory name, not a file), return it with
    the trailing slash. Otherwise return ''.
    """
    prefixes = set()
    for m in members:
        parts = m.name.replace("\\", "/").split("/")
        if parts:
            prefixes.add(parts[0])
    if len(prefixes) == 1:
        candidate = prefixes.pop()
        # Only strip if it looks like a directory name (no extension)
        if "." not in candidate:
            return candidate + "/"
    return ""


def _prune_old_versions(versions_dir: str, keep: int) -> None:
    """Delete oldest snapshots beyond the keep limit."""
    if keep <= 0:
        return
    entries = []
    for name in os.listdir(versions_dir):
        full = os.path.join(versions_dir, name)
        try:
            entries.append((os.stat(full).st_mtime, full))
        except OSError:
            pass

    entries.sort(reverse=True)   # newest first
    for _, path in entries[keep:]:
        try:
            if os.path.isdir(path):
                import shutil
                shutil.rmtree(path)
            else:
                os.remove(path)
        except OSError:
            pass

# ── GPLv3 interactive notice ──────────────────────────────────────────────────

_GPLv3_WARRANTY = (
    "THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY\n"
    "APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT\n"
    'HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM \"AS IS\" WITHOUT\n'
    "WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT\n"
    "LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A\n"
    "PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE\n"
    "OF THE PROGRAM IS WITH YOU.  (GPL-3.0-or-later §15)"
)

_GPLv3_CONDITIONS = (
    "You may convey verbatim copies of the Program's source code as you\n"
    "receive it, in any medium, provided that you conspicuously and\n"
    "appropriately publish on each copy an appropriate copyright notice and\n"
    "disclaimer of warranty. (See GPL-3.0 §4-6 for full conditions.)\n"
    "Full license: <https://www.gnu.org/licenses/gpl-3.0.html>"
)


def gplv3_notice():
    """Print the short GPLv3 startup notice. Call this at program startup."""
    print("S-IDE  Copyright (C) 2026  N0V4-N3XU5")
    print("This program comes with ABSOLUTELY NO WARRANTY; for details type 'show w'.")
    print("This is free software, and you are welcome to redistribute it")
    print("under certain conditions; type 'show c' for details.")


def gplv3_handle(cmd: str) -> bool:
    """
    Check whether *cmd* is a GPLv3 license command and handle it.
    Returns True if the command was consumed (caller should skip normal processing).
    """
    match cmd.strip().lower():
        case "show w":
            print(_GPLv3_WARRANTY)
            return True
        case "show c":
            print(_GPLv3_CONDITIONS)
            return True
    return False

# ─────────────────────────────────────────────────────────────────────────────
